cpp/src/arrow/util/decimal.cc

```diff
@@ -176,37 +177,46 @@ static constexpr int64_t kPowersOfTen[kInt64DecimalDigits + 1] = {1LL,
                                                                   100000000000000000LL,
                                                                   1000000000000000000LL};
 
-static void StringToInteger(const std::string& str, Decimal128* out) {
+// Iterates over data and for each group of kInt64DecimalDigits multiple out by
+// the appropriate power of 10 necessary to add source parsed as uint64 and
+// then adds the parsed value of source.
+static inline void ShiftAndAdd(const char* data, size_t length, Decimal128* out) {
+  static internal::StringConverter<Int64Type> converter;
+  for (size_t posn = 0; posn < length;) {
+    const size_t group_size = std::min(kInt64DecimalDigits, length - posn);
+    const int64_t multiple = kPowersOfTen[group_size];
+    int64_t chunk = 0;
+    ARROW_CHECK(converter(data + posn, group_size, &chunk));
```

Comment:
I think we should probably return status here.  I think the exponent could overflow (I think this was a potential bug in the old code as well).